### PR TITLE
performance: add case "Virio with qemu vhost Back end in DPDK"

### PR DESCRIPTION
**Pull request by quanwenli**
id: 1845797

Signed-off-by: Wenli Quan <wquan@redhat.com>

### CODE REVIEW

**Comment by quanwenli** [comment]
(1/1) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.virtio_net_dpdk.q35: PASS (345.08 s)
outputttt=True
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Comment by quanwenli** [comment]
@vivianQizhu, Could you help review it, thanks.

File: qemu/tests/virtio_net_dpdk.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+    def _pin_vm_threads(node):
+        """
+        pin guest vcpu and vhost threads to cpus of a numa node repectively
+
+        :param node: which numa node to pin
+        """
+        if node:
+            if not isinstance(node, utils_misc.NumaNode):
+                node = utils_misc.NumaNode(int(node))
+            utils_test.qemu.pin_vm_threads(vm, node)
```
Comment (vivianQizhu) [inline qemu/tests/virtio_net_dpdk.py:57]: Skip this step if node is not provided?

File: qemu/tests/virtio_net_dpdk.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        3) Copy testpmd script to guest
+
+        """
+        error_context.context("Setup env for guest")
+
+        # setup hugepages
+        session.cmd(params.get("env_hugepages_cmd"), ignore_all_errors=True)
+
+        # install tuned-profiles-cpu-partitioning dpdk packages
+        cmd = 'yum install -y %s' % params.get("env_pkg")
```
Comment (vivianQizhu) [inline qemu/tests/virtio_net_dpdk.py:73]: Prefer to use `dnf`, and please use params[''] if the value can not be `None`

File: qemu/tests/virtio_net_dpdk.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+            for i in nic_driver:
+                if i == "Virtio":
+                    nic_pci_1 = "0000:%s" % session.cmd(
+                        "lspci |awk '/%s network/ {print $1}'" % i).strip()
+                    cmd_str = "dpdk-devbind --bind=vfio-pci %s" % nic_pci_1
+                if i == "82599":
+                    nic_pci_2 = "0000:%s" % session.cmd(
+                        "lspci |awk '/%s/ {print $1}'" % i).strip()
+                    cmd_str = "dpdk-devbind --bind=vfio-pci %s" % nic_pci_2
+                session.cmd_output(cmd_str)
```
Comment (vivianQizhu) [inline qemu/tests/virtio_net_dpdk.py:110]: Do we need to restrict nic type in cfg file if we don't handle other types?

File: qemu/tests/virtio_net_dpdk.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    macvtap_mac = vm.get_mac_address(1)
+    vfio_mac = vm.get_mac_address(2)
+
+    # get parameter from dictionary
+    category = params.get("category")
+    pkt_size = params.get("pkt_size")
+    kvm_ver_chk_cmd = params.get("kvm_ver_chk_cmd")
+    guest_ver_cmd = params["guest_ver_cmd"]
+    guest_dpdk_cmd = params["guest_dpdk_cmd"]
+    record_list = params.get("record_list")
```
Comment (vivianQizhu) [inline qemu/tests/virtio_net_dpdk.py:181]: Can be empty?

File: qemu/tests/virtio_net_dpdk.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        cmd += " %s %s %s %s %s %s %s %s > /tmp/testpmd.log" % (
+            exec_file, nic1_driver, nic2_driver,
+            nic_pci_1, nic_pci_2, cores, queues, running_time)
+        session.cmd_output(cmd)
+
+    moongen_thread = threading.Thread(
+        target=start_moongen, args=(generator1, mac, port_id, running_time))
+    moongen_thread.start()
+
+    if utils_misc.wait_for(lambda: run_moongen_up(generator2), 30, 1, 1,
```
Comment (vivianQizhu) [inline qemu/tests/virtio_net_dpdk.py:314]: Default step is 1 so you don't have to specify it.

**Comment by quanwenli** [comment]
@@vivianQizhu, I have updated all your comments, but I am not sure why I can not reply with your each comments. please help review.

**Comment by quanwenli** [comment]
JOB LOG    : /root/avocado/job-results/job-2020-06-19T06.07-a0e2fe1/job.log
 (1/1) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.2.1.x86_64.io-github-autotest-qemu.virtio_net_dpdk.q35: PASS (164.07 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2020-06-19T06.07-a0e2fe1/results.html

File: qemu/deps/performance/start_testpmd.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    :param data: data to be transformed into text
+    :param encoding: encoding of the data (only used when decoding
+                     is necessary)
+    :type data: either bytes or other data that will be returned
+                unchanged
+    """
+    if isinstance(data, bytes):
+        if encoding is None:
+            encoding = ENCODING
+        return data.decode(ENCODING)
```
Comment (vivianQizhu) [inline qemu/deps/performance/start_testpmd.py:146]: Here should be `data.decode(encoding)`? Otherwise why you define `encoding`?

Comment (vivianQizhu) [reply qemu/deps/performance/start_testpmd.py:146]: If `return data.decode(ENCODING)`, then Line 143-145 is useless so should be removed.

**Comment by quanwenli** [comment]
@vivianQizhu, I have updated your latest comment about "return data.decode(ENCODING) is enough.", please help review.

File: qemu/deps/performance/start_testpmd.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+    :param data: data to be transformed into text
+    :param encoding: encoding of the data (only used when decoding
+                     is necessary)
+    :type data: either bytes or other data that will be returned
+                unchanged
+    """
+    if is_bytes(data):
+        if encoding is None:
+            encoding = ENCODING
```
[the remark this answers is not in the document]
Comment (vivianQizhu) [reply qemu/deps/performance/start_testpmd.py:145]: The definition of `encoding` is meaningless since you use `ENCODING` directly in the below.

**Pull request merged**